### PR TITLE
Benchmark Block struct serialization code

### LINKED ISSUE #2019: Benchmark Zebra's block serialization code
**Is your feature request related to a problem? Please describe.**

Flowing from benchmarking message serialization in #1775, we want some block serialization becnhmarks to make sure that the changes actually improved performance.

**Describe the solution you'd like**

Add benchmarks for block serialisation and deserialisation.

**Describe alternatives you've considered**

Don't benchmark: we continue to guess about performance, and sometimes we'll get it wrong

### PR DESCRIPTION
**Pull request by fanatid**
## Motivation

For resolving #1774 there should be a way to determine that improvement was introduced. Benchmark is a good way to see performance changes.
I do not think that it's required to benchmark all structs which implement `ZcashSerialize`. `Block` uses a lot (all?) of structs internally, so only `Block` should be enough.

## Solution

Add a simple benchmark for the biggest block from `zebra-test` created with [criterion](https://crates.io/crates/criterion).

## Related Issues

Should close https://github.com/ZcashFoundation/zebra/issues/2019.

## Follow Up Work

Create tracking issue for adding benchmarks to CI once this PR is merged.

### CODE REVIEW

**Comment by dconnolly** [comment]
Thanks for this PR! This will benchmark serialization of the largest block test vector we have, but not the messages we support yet, so we should create another issue and resolve that with this PR vs #1775. 

And yes we should run all our benchmarks as part of CI, some of our dependencies (ed25519-zebra, redjubjub, bellman) have benchmarks for the batch verification of those signatures/proofs that we may want to run ourselves, or, repurpose them on top of the async services that leverage those batching implementations, and run those instead.

File: zebra-chain/benches/block.rs
Comment (dconnolly) [inline zebra-chain/benches/block.rs]: @teor2345 do you think this largest block as the serialization benchmark is sufficient? I know our test vectors are not comprehensive in terms of the possible variety of outlier blocks (all transparent transactions, all shielded transactions of different types, just coinbase, etc) but that's more a test coverage/long tail issue than a benchmark issue.

Comment (teor2345) [reply zebra-chain/benches/block.rs]: It depends on the purpose of the benchmark.

if it was security-related (for example, testing denial of service conditions), we'd want to try every different combination, probably using randomised testing.

For now I suggest we test:
- the largest public block
- the synthetic big blocks in our existing unit tests: https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/tests/generate.rs#L18

And then we can open a follow-up ticket for different transaction types. (We should probably test a big orchard transaction as part of the NU5 changes, I'll add it to our list in that ticket.)

Comment (teor2345) [reply zebra-chain/benches/block.rs]: > And then we can open a follow-up ticket for different transaction types. (We should probably test a big orchard transaction as part of the NU5 changes, I'll add it to our list in that ticket.)

I created #2030 for sprout and sapling, and updated #1979 for orchard.

File: zebra-chain/src/block/tests.rs
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: The second `mod generate` is redundant and causes a clippy warning:
```suggestion
#[cfg(any(test, feature = "bench"))]
pub mod generate;
```

File: zebra-chain/benches/block.rs
Comment (teor2345) [reply zebra-chain/benches/block.rs]: Sorry I wasn't very clear here, let's test both:
- large_single_transaction_block
- large_multi_transaction_block (already implemented)

The performance of a large number of transactions vs a large number of inputs could be quite different.

Comment (fanatid) [reply zebra-chain/benches/block.rs]: Added.

File: Cargo.toml
Comment (teor2345) [inline Cargo.toml:37]: Since this PR doesn't change zcash_script or RocksDB, I'm guessing the lexical-core change might actually be part of #2026.

But there's no need to move it, unless there's merge conflicts, or you think we might have to revert the M1 PR.

File: zebra-chain/benches/block.rs
Comment (teor2345) [inline zebra-chain/benches/block.rs:26]: 👍

File: zebra-chain/benches/block.rs
Comment (teor2345) [inline zebra-chain/benches/block.rs]: Looks like there was an incomplete search and replace here.

I suggest using `fastmod` for interactive search and replace, it's a good way to make sure you don't miss anything.
```suggestion
            BenchmarkId::new("zcash_deserialize", name),
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
Let's fix the hard-coded block name, and then we're good to merge.

You can also move the lexical-core change to #2026 if you want, but I don't want to block on that.

File: Cargo.toml
Comment (teor2345) [reply Cargo.toml:37]: Resolved by #2026

**Review by teor2345 — APPROVED** [review]
Once CI passes, let's auto-merge this PR.

**Comment by fanatid** [comment]
😮 @teor2345 how you pushed 1761a5b to my fork?

**Comment by teor2345** [comment]
> 😮 @teor2345 how you pushed [1761a5b](https://github.com/ZcashFoundation/zebra/commit/1761a5b543c66cd3dc51e91af8ac3ce5b7f230f6) to my fork?

GitHub has an "allow maintainer updates" setting on PRs. It's on by default, and it lets maintainers push to your PR branch.

Sorry I should have asked first 🤭

(I always ask for permission for big changes.)

**Comment by fanatid** [comment]
> > 😮 @teor2345 how you pushed 1761a5b to my fork?

> GitHub has an "allow maintainer updates" setting on PRs. It's on by default, and it lets maintainers push to your PR branch.
> Sorry I should have asked first 🤭
> (I always ask for permission for big changes.)

wow, I thought this checkbox only for editing initial message. TIL 👍

**Pull request merged**
